File: src/ai_dm/rules/cover.py
"""Cover — SRD 5.2.

Single source of truth: ``assets/srd5_2/core/cover.json``.

Cover grants a bonus to AC and Dexterity saving throws against effects
that originate outside the cover. Total cover blocks the line of
effect entirely.
"""
from __future__ import annotations

from typing import Literal

from ai_dm.rules.srd_core import load

_DATA = load("cover")
CoverKey = Literal["none", "half", "three_quarters", "total"]

_LEVELS = {rec["key"]: rec for rec in _DATA["levels"]}
COVER_KEYS: tuple[str, ...] = tuple(_LEVELS.keys())


def _level(key: str) -> dict:
    return _LEVELS.get(key, _LEVELS["none"])


def ac_bonus(cover: str) -> int:
    val = _level(cover).get("ac")
    return 0 if val is None else int(val)


def dex_save_bonus(cover: str) -> int:
    val = _level(cover).get("save")
    return 0 if val is None else int(val)


def blocks(cover: str) -> bool:
    """True if the cover prevents the attack/effect entirely (total cover)."""
    return bool(_level(cover).get("blocks", False))


def apply_to_target_ac(target_ac: int, *, cover: str) -> int:
    """Effective AC the attacker must beat after cover."""
    return int(target_ac) + ac_bonus(cover)


__all__ = [
    "COVER_KEYS",
    "CoverKey",
    "ac_bonus",
    "apply_to_target_ac",
    "blocks",
    "dex_save_bonus",
]

File: foundry/module/scripts/srd/cover.js
/** Cover — mirrors `ai_dm.rules.cover`. */
import { srd } from "./core_loader.js";

export const COVER = ["none", "half", "three_quarters", "total"];

function _level(key) {
    const data = srd()?.cover?.levels ?? [];
    return data.find((l) => l.key === key) ?? data.find((l) => l.key === "none") ?? {};
}

export function acBonus(cover) {
    const v = _level(cover).ac;
    return v == null ? 0 : Number(v);
}

export function dexSaveBonus(cover) {
    const v = _level(cover).save;
    return v == null ? 0 : Number(v);
}

export function blocks(cover) {
    return Boolean(_level(cover).blocks);
}

export function applyToTargetAc(targetAc, { cover = "none" } = {}) {
    return Number(targetAc) + acBonus(cover);
}

